primitives/transaction/src/historic_transaction.rs

```diff
@@ -159,15 +159,33 @@ impl HistoricTransaction {
         }
     }
 
-    /// Returns the hash of the underlying transaction/inherent. For reward inherents we return the
-    /// hash of the corresponding reward transaction. This results into an unique hash for the
-    /// reward inherents (which wouldn't happen otherwise) and allows front-end to fetch rewards by
-    /// their transaction hash.
+    /// Returns the hash based on the underlying transaction/event.
     pub fn tx_hash(&self) -> Blake2bHash {
         match &self.data {
             HistoricTransactionData::Basic(tx) => tx.hash(),
-            // TODO: check for freedom of hash prefixes
-            _ => self.data.hash(),
+            _ => {
+                let block_height = if !matches!(self.data, HistoricTransactionData::Equivocation(_))
+                {
+                    self.block_number
+                } else {
+                    // Equivocations are not distinguished by block height.
+                    0
+                };
+                let mut hasher = Blake2bHasher::new();
+                // Ensure that the data here doesn't form a valid transaction
+                // by adhering to the same serialization structure as
+                // `Transaction::serialize_content`.
+                hasher.write_all(b"\x00\x00").unwrap(); // recipient data len
+                {} // recipient data
+                hasher.write_all(b"\x00\x00\x00\x00\x00\x00\x00\x00").unwrap(); // sender
+                hasher.write_all(b"\x00\x00\x00\x00\x00\x00\x00\x00").unwrap(); // sender
+                hasher.write_all(b"\x00\x00\x00\x00").unwrap(); // sender
+                hasher.write_all(b"\xff").unwrap(); // sender_type (HistoricTransaction = 0xff)
+                self.network_id.serialize_to_writer(&mut hasher).unwrap();
+                block_height.serialize_to_writer(&mut hasher).unwrap();
+                self.data.serialize_to_writer(&mut hasher).unwrap();
+                hasher.finish()
```

Comment:
I don't understand the comment. Do you want to adhere to a transaction format, or don't? Because you write you don't, but then start exactly like a transaction would. But then after the sender type serialize completely different. So it is _not_ a transaction serialization, in which case the static data length and sender stuff make no sense and can be left out.

```suggestion
                let mut hasher = Blake2bHasher::new();
                block_height.serialize_to_writer(&mut hasher).unwrap();
                self.network_id.serialize_to_writer(&mut hasher).unwrap();
                self.data.serialize_to_writer(&mut hasher).unwrap();
                hasher.finish()
```